### PR TITLE
Lora

### PR DESCRIPTION
**Pull request by ZidongLiu**
# Summary

- [x] Train 125m model lora with first order
- [x] Find proper lora parameters, in opt-125m, lr = 1e-3(ZO all parameters is 1e-5). We can assume lr is 100 times larger than our usual cases, @RogerRogerUSC @BichengYing just FYI
- [x] Train model using cezo_fl
- [x] Make float16 and float32 work

# Results
float32 works
image

float16 works
image

### CODE REVIEW

File: shared/model_helpers.py [hunk 1]
```diff
@@ -57,9 +59,18 @@ def load_model_and_optimizer(optimizer, model, model_path):
...
     return model, optimizer
+
+
+def get_trainable_model_parameters(
+    model: torch.nn.Module,
+) -> Callable[[], Iterator[torch.nn.Parameter]]:
+    def generator():
+        return filter(lambda x: x.requires_grad, model.parameters())
+
+    return generator
```
Comment (BichengYing) [inline shared/model_helpers.py:76]: I think the better implementation should be
```python
for param in model.parameters():
    if param.requires_grad:
        yield param
```

In this way, you can just pass `get_trainable_model_parameters(model)` into the functions

File: gradient_estimators/random_gradient_estimator.py [hunk 2]
```diff
@@ -1,7 +1,8 @@
 import torch
 from torch.nn import Parameter
 from typing import Callable, Iterator, TypeAlias, Literal, Sequence
-import transformers
+from transformers.models.opt.modeling_opt import OPTForCausalLM
+from peft import PeftModel
```
[the remark this answers is not in the document]
Comment (ZidongLiu) [reply gradient_estimators/random_gradient_estimator.py:5]: I already included it in the env beforehand

File: decomfl_main.py [hunk 3]
```diff
@@ -71,13 +77,24 @@ def prepare_settings_underseed(args, device):
         model_name = SUPPORTED_LLM[large_model]
         model = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch_dtype).to(device)
         model.model_name = large_model
+        if args.lora:
+            # this step initialize lora parameters, which should be under control of seed
+            lora_config = LoraConfig(
+                r=args.lora_r, lora_alpha=args.lora_alpha, target_modules=["q_proj", "v_proj"]
+            )
+            model = get_peft_model(model, lora_config).to(torch_dtype)
```
[the remark this answers is not in the document]
Comment (ZidongLiu) [reply decomfl_main.py:85]: super easy to use

**Comment by ZidongLiu** [comment]
Lora still works when (lr = 1e-3)
image

File: shared/model_helpers.py (hunk 1 above)
Comment (ZidongLiu) [reply shared/model_helpers.py]: resolved and added a unit test

**Comment by ZidongLiu** [comment]
zp_rge_main still works

image

**Review by BichengYing — APPROVED** [review]
LGTM!

**Pull request merged**